hips/hip-00NN.md

```diff
@@ -0,0 +1,101 @@
+type: "feature"
+status: "draft"
+---
+
+## Abstract
+
+Currently, Helm supports installing dependencies from various registries, however that requires the charts to be published there. There are use cases which are cumbersome when registries are involved, for example:
+
+- private charts for organizations/individuals who do not have the capacity to maintain the infrastructure required to run a private registry
+- testing work-in-progress charts in their umbrella charts
+
+## Motivation
+
+There are existing ways to achieve installation of charts without using a registry, however they are not very user-friendly and require additional tooling.
+
+- A Helm chart repository is effectively an `index.yaml` with links to downloads. Maintaining such a repository does create a burden of scripts and automation (e.g. Github Actions). This is not always feasible for smaller projects. It also does not really offer an obvious and readily available way of testing pre-releases.
+- For the testing use cases, charts can be packaged using `helm package`, however this does introduce manual steps and requires extra work to replicate in CI/CD scenarios.
+- There are several plugins available to solve this problem ([`aslafy-z/helm-git`](https://github.com/aslafy-z/helm-git), [`diwakar-s-maurya/helm-git`](https://github.com/diwakar-s-maurya/helm-git), [`sagansystems/helm-github`](https://github.com/sagansystems/helm-github)), however using plugins requires additional setup, which may not always be feasible (esp. in more complex team structures and cluster setups with advanced tooling, e.g. ArgoCD). An additional drawback is that the `Chart.yaml` does not provide a way to specify the plugin requirements, which leaves it up to the consumer to figure this out.
+
+## Rationale
+
+Installing dependencies from git is an established pattern in other ecosystems even when they are registry-based, e.g. npm (Node.js), pipenv (Python), bundler (Gem) have this option - it would make sense to have the behavior replicated in Helm.
+
+At least the npm and the pip ecosystems have already established a syntax of `vcs+protocol` for defining the dependency source, so it should be familiar to some users.
+
+One alternative to consider would be to exclude defining the vcs/protocol, similar to what Go does, esp. given that Helm is built using Go. This does limit the flexibility somewhat - while Go does allow adding a VCS qualifier at the end of the URL (allowing future support for other VCSs), it does not allow specifying the protocol, which means that the users might have to override the default protocol in their VCS configuration.
+
+## Specification
+
+The `Chart.yaml` should support the following format for `dependencies`:
+
+```
+dependencies:
+- name: "<dependency name>"
+  repository: "<protocol>://[<user>[:<password>]@]<hostname>[:<port>][:][/]<path>"
+  version: "<commit-ish>"
+```
+where:
+- `<protocol>` is one of `git`, `git+ssh`, `git+http`, `git+https`, or `git+file`.
```

Comment:
I wonder if we could generally specify `git[+<subprotocol>]://` should be handled by git? Rather than explicitly listing the above set of protocols?

I don't think an explicit list is necessary bad. It's not like new transports will be added often. But similarly curious if that could/should be avoided.

Reply:
Reworded to take `git[+<subprotocol>]` into account, but still listing the protocols that git supports as an example. I figure the rest should really be tackled via documentation? Added a note under "How to teach this": https://github.com/helm/community/pull/321/commits/0f6cd2fc9795e52c2c3f33adf769e9fee3374c4c